### PR TITLE
[Model] Bert Embedding Model

### PR DESCRIPTION
**Pull request by laishzh**
Implement Bert Embedding Model

This PR implements the Bert Embedding Model which discussed in #5179.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/models/bert_embedding.py [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+                          attn_metadata=attn_metadata)
+
+    def pooler(
+        self,
+        hidden_states: torch.Tensor,
+        pooling_metadata: PoolingMetadata,
+    ) -> Optional[PoolerOutput]:
+        return self._pooler(hidden_states, pooling_metadata)
+
+    def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/models/bert_embedding.py:63]: Can you move this function down to the bottom of the file

File: vllm/model_executor/models/bert_embedding.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+        return output
+
+
+class BertEmbedding(nn.Module):
+
+    def __init__(self, config: BertConfig):
+        super().__init__()
+        self.size = config.hidden_size
+
+        self.word_embeddings = nn.Embedding(config.vocab_size,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/models/bert_embedding.py:149]: Could you look into using `VocabParallelEmbedding` from our parallel layers?

Using the `nn.Embedding` will not work with tensor parallelism

File: vllm/model_executor/models/bert_embedding.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+                                            config.hidden_size,
+                                            padding_idx=config.pad_token_id)
+        self.position_embeddings = nn.Embedding(config.max_position_embeddings,
+                                                config.hidden_size)
+        self.token_type_embeddings = nn.Embedding(config.type_vocab_size,
+                                                  config.hidden_size)
+        self.LayerNorm = nn.LayerNorm(config.hidden_size,
+                                      eps=config.layer_norm_eps)
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/models/bert_embedding.py:158]: Do we need dropout for inference?

File: vllm/model_executor/models/bert_embedding.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+        self.scaling = self.head_dim**-0.5
+        self.qkv_proj = QKVParallelLinear(
+            hidden_size=self.hidden_size,
+            head_size=self.head_dim,
+            total_num_heads=self.total_num_heads,
+            total_num_kv_heads=self.total_num_kv_heads,
+            bias=True,
+            quant_config=quant_config)
+
+        self.attn = Attention(
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/models/bert_embedding.py:332]: We need to use bidirectional attention here rather than causal attention

**Comment by robertgshaw2-neuralmagic** [comment]
Thanks! You're on the right track here

File: vllm/model_executor/models/bert_embedding.py [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+        self.size = config.hidden_size
+
+        self.word_embeddings = nn.Embedding(config.vocab_size,
+                                            config.hidden_size,
+                                            padding_idx=config.pad_token_id)
+        self.position_embeddings = nn.Embedding(config.max_position_embeddings,
+                                                config.hidden_size)
+        self.token_type_embeddings = nn.Embedding(config.type_vocab_size,
+                                                  config.hidden_size)
+        self.LayerNorm = nn.LayerNorm(config.hidden_size,
```
Comment (mgoin) [inline vllm/model_executor/models/bert_embedding.py:156]: Follow naming standards for layernorm
```suggestion
        self.layernorm = nn.LayerNorm(config.hidden_size,
```

File: vllm/model_executor/models/bert_embedding.py [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+
+
+class BertSelfOutput(nn.Module):
+
+    def __init__(self, config: BertConfig):
+        super(BertSelfOutput, self).__init__()
+        self.dense = nn.Linear(config.hidden_size, config.hidden_size)
+        self.LayerNorm = nn.LayerNorm(config.hidden_size,
+                                      eps=config.layer_norm_eps)
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
```
Comment (mgoin) [inline vllm/model_executor/models/bert_embedding.py:360]: ditto on layernorm and dropout

File: vllm/model_executor/models/bert_embedding.py [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+
+
+class BertOutput(nn.Module):
+
+    def __init__(self, config: BertConfig):
+        super().__init__()
+        self.dense = nn.Linear(config.intermediate_size, config.hidden_size)
+        self.LayerNorm = nn.LayerNorm(config.hidden_size,
+                                      eps=config.layer_norm_eps)
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
```
Comment (mgoin) [inline vllm/model_executor/models/bert_embedding.py:389]: ditto on layernorm and dropout

**Comment by laishzh** [comment]
> Thanks! You're on the right track here

Wow! Really appreciate your guidance.

File: vllm/model_executor/models/bert_embedding.py [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+
+                # Skip loading extra bias for GPTQ models.
+                if name.endswith(".bias") and name not in params_dict:
+                    continue
+
+                param = params_dict[name]
+                weight_loader = param.weight_loader
+                weight_loader(param, loaded_weight, shard_id)
+                break
+            else:
```
[the remark this answers is not in the document]
Comment (laishzh) [reply vllm/model_executor/models/bert_embedding.py]: Yes, I also felt confused at the first time. I rewote this part just now. It's supposed to be easier to understand. Have any further suggestions?

File: vllm/model_executor/models/bert_embedding.py (hunk 1 above)
Comment (laishzh) [reply vllm/model_executor/models/bert_embedding.py:86]: OK, leave a TODO here. I will move this class to the bottom later.

File: vllm/model_executor/models/bert_embedding.py (hunk 2 above)
Comment (laishzh) [reply vllm/model_executor/models/bert_embedding.py]: OK, I noticed this feature before. Already updated.

File: vllm/model_executor/models/bert_embedding.py (hunk 5 above)
Comment (laishzh) [reply vllm/model_executor/models/bert_embedding.py]: Followed the suggestion, and added renaming of LayerNorm parameter when loading weights.

File: vllm/model_executor/models/bert_embedding.py (hunk 3 above)
Comment (laishzh) [reply vllm/model_executor/models/bert_embedding.py]: All dropouts have been removed.

File: vllm/model_executor/models/bert_embedding.py (hunk 4 above)
Comment (laishzh) [reply vllm/model_executor/models/bert_embedding.py:344]: Do you mean to implement a new type of attention in vllm? Not sure if there is a way to use the current attention implementation with different parameters.

Comment (laishzh) [reply vllm/model_executor/models/bert_embedding.py:344]: @robertgshaw2-neuralmagic hi, I'd like to hear your thoughts on this point. The framework of BERT model is almost completed. But the BertSelfAttention output differs from transformers. After diving into the Attention implementation, there are massive changes of Attention are needed to use bidirectional attention.
I also find some related discussions(https://github.com/vllm-project/vllm/pull/3117#issuecomment-2159049996). I think this PR is depended on #4942. What's your opinions?

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/models/bert_embedding.py:344]: @laishzh sorry missed this note!

Yes. Pull in the `BertSelfAttention` from that PR. #4942 should land this week

**Comment by laishzh** [comment]
@robertgshaw2-neuralmagic @mgoin Hi, just update this work. 
Because the BertEmbeddingModel is encoder-only architecture. I refactored the EmbeddingModelRunner as child class of EncoderDecoderModelRunner to reuse the encoder-decoder framework.

File: vllm/worker/embedding_model_runner.py [hunk 9]
```diff
@@ -137,14 +149,34 @@ def prepare_model_input(
...
+            model_input,
+            attn_metadata=attn_metadata,
+            encoder_input_tokens=encoder_input_tokens_tensor,
+            encoder_input_positions=encoder_input_positions_tensor,
+        )
+
         # Prepare PoolingMetadata.
-        assert model_input.seq_lens is not None
+        seq_lens = model_input.seq_lens\
+            if not self.model_config.is_encoder_model \
```
Comment (laishzh) [inline vllm/worker/embedding_model_runner.py:174]: @robertgshaw2-neuralmagic @mgoin Here still have a question. How to determine precisely which model is encoder model or decoder model. As an example, there is lake of fields `is_decoder_model` or `is_encoder_model` in `config.json` of `Mistral Model` which is implemented before. Here is the link: [config.json of Mistral](https://huggingface.co/intfloat/e5-mistral-7b-instruct/blob/main/config.json).

**Comment by maxdebayser** [comment]
Hi @laishzh, I'm working on another PR that is based on yours. In case it helps, I've solved the latest merge conflicts of your branch with `main` here: https://github.com/maxdebayser/vllm/tree/bert

**Comment by maxdebayser** [comment]
@laishzh @robertgshaw2-neuralmagic , I've solved that lastest merge conflicts here: https://github.com/maxdebayser/vllm/tree/bert

**Comment by laishzh** [comment]
@robertgshaw2-neuralmagic Please draw your attention. I just revert the change to EmbeddingModelBlockManager. And there is still leaving a problem of how to distinguish the model arch(Encoder-Only, or Decoder-Only model), which is to determine the seq_len when pooling(https://github.com/vllm-project/vllm/pull/5447#discussion_r1750487428). Please let me know if any changes are needed.

**Comment by robertgshaw2-neuralmagic** [comment]
Hey @laishzh - there were a few issues with this PR. Specifically, it uses the Encoder-Decoder pathway and the BertModel is not implemented properly (loading is not canonical, tensor parallelism does not work, and the pooling logic is not correct)

I need get this landed ASAP, so I finished off the PR here. https://github.com/vllm-project/vllm/pull/9056

I will added you as a co-author.

**Comment by laishzh** [comment]
> Hey @laishzh - there were a few issues with this PR. Specifically, it uses the Encoder-Decoder pathway and the BertModel is not implemented properly (loading is not canonical, tensor parallelism does not work, and the pooling logic is not correct)
> 
> I need get this landed ASAP, so I finished off the PR here. #9056
> 
> I will added you as a co-author.

I'm OK. Very glad to see this feature could be supported in vllm. I will delve into those points that you mentioned. Thanks for your work!

**Comment by DarkLight1337** [comment]
Closing as superseded by #9056
